contrib/dracut/90zfs/mount-zfs.sh.in

```diff
@@ -61,10 +61,11 @@ if import_pool "${ZFS_POOL}" ; then
 		# if the root dataset has encryption enabled
 		if $(zfs list -H -o encryption "${ZFS_DATASET}" | grep -q -v off); then
 			# figure out where the root dataset has its key, the keylocation should not be none
+			ZFS_CRYPTPARENT="$ZFS_DATASET"
 			while true; do
-				if [[ $(zfs list -H -o keylocation "${ZFS_DATASET}") == 'none' ]]; then
-					ZFS_DATASET=$(echo -n "${ZFS_DATASET}" | awk 'BEGIN{FS=OFS="/"}{NF--; print}')
-					if [[ "${ZFS_DATASET}" == '' ]]; then
+				if [[ $(zfs list -H -o keylocation "${ZFS_CRYPTPARENT}") == 'none' ]]; then
+					ZFS_CRYPTPARENT=$(echo -n "${ZFS_CRYPTPARENT}" | awk 'BEGIN{FS=OFS="/"}{NF--; print}')
+					if [[ "${ZFS_CRYPTPARENT}" == '' ]]; then
```

Comment:
Not a bash expert or anything here, but it looks like you are trying to find the encryption root for the dataset by iterating through its parents.  Unfortunately, the relationship between datasets and their encryption roots is a bit complicated. For instance, clones always share an encryption key with their origins, which may or may not actually be related to them as a parent (ex. `pool/clone` as a clone of `pool/crypt` ). A much more reliable way to do this would be to use the `encryptionroot` property, which was specifically added for this kind of problem.

Also (forgive me, but I'm also not an expert in different init systems or this script in particular) why would we not just use `zpool import -l` here? It will attempt to load all keys as the pool is imported.

Reply:
I believe `zpool import -l` was my original suggestion before the init scripts were updated for encryption, but it was decided that we shouldn't try to load all the keys, only those needed for booting.  @prometheanfire and @kpande can probably back me up on that a little.

Reply:
yes, that's the reason behind zpool import -l not being used.

I'll look at the encrypyionroot option, I probably just missed it when going over the dataset properties...